tests/Doctrine/Tests/DBAL/Query/Expression/CompositeExpressionTest.php

```diff
@@ -77,4 +77,12 @@ public function provideDataForConvertToString()
             ),
         );
     }
+
+    public function testAddPreventsAddingEmptyCompositeExpressionObjects()
+    {
+        $emptyExpr = new CompositeExpression(CompositeExpression::TYPE_AND, []);
+        $expr = new CompositeExpression(CompositeExpression::TYPE_AND, [$emptyExpr]);
+
+        $this->assertEquals(0, count($expr));
```

Comment:
No, it was just wishful thinking: if we didn't allow mutation of expressions, this problem would not exist in first place.

Anyway, the fix doesn't seem correct to me: I think the correct way of dealing with the problem is to remove unnecessary composite expression building within `__toString`, which is called at the very end of the entire procedure. This way, even if state changed in-between, we'd still get consistent behavior.